File: src/200.number-of-islands/solution_dfs-recursive.ts
/*
 * @lc app=leetcode.cn id=200 lang=typescript
 *
 * https://leetcode-cn.com/problems/number-of-islands/
 */

// @lc code=start

// grid is not empty
// row, col are valid
function clearIslandDfsRecursive(grid: string[][], row: number, col: number) {
  const rowNum = grid.length;
  const colNum = grid[0].length;

  // reset current element
  grid[row][col] = '0';

  // up
  if (row - 1 >= 0 && grid[row - 1][col] === '1') {
    grid[row - 1][col] = '0';
    clearIslandDfsRecursive(grid, row - 1, col);
  }
  // down
  if (row + 1 < rowNum && grid[row + 1][col] === '1') {
    grid[row + 1][col] = '0';
    clearIslandDfsRecursive(grid, row + 1, col);
  }
  // left
  if (col - 1 >= 0 && grid[row][col - 1] === '1') {
    grid[row][col - 1] = '0';
    clearIslandDfsRecursive(grid, row, col - 1);
  }
  // right
  if (col + 1 < colNum && grid[row][col + 1] === '1') {
    grid[row][col + 1] = '0';
    clearIslandDfsRecursive(grid, row, col + 1);
  }
}

function numIslands(grid: string[][]): number {
  // grid is empty
  if (grid.length === 0 || grid[0].length === 0) {
    return 0;
  }

  const rowNum = grid.length;
  const colNum = grid[0].length;
  let islandNum = 0;

  for (let row = 0; row < rowNum; ++row) {
    for (let col = 0; col < colNum; ++col) {
      if (grid[row][col] === '1') {
        islandNum += 1;
        clearIslandDfsRecursive(grid, row, col);
      }
    }
  }

  return islandNum;
}

// @lc code=end

export default numIslands;

File: src/200.number-of-islands/solution_dfs-recursive.cpp
/*
 * @lc app=leetcode.cn id=200 lang=cpp
 *
 * https://leetcode-cn.com/problems/number-of-islands/
 */

#include <stack>
#include <utility>
#include <vector>

using namespace std;

// @lc code=start

class Solution {
 private:
  // grid is not empty
  // row, col are valid
  void clearIslandDfsRecursive(vector<vector<char>> &grid, int row, int col) {
    int row_num = grid.size();
    int col_num = grid[0].size();

    // clear current element
    grid[row][col] = '0';

    // up
    if (row - 1 >= 0 && grid[row - 1][col] == '1') {
      grid[row - 1][col] = '0';
      clearIslandDfsRecursive(grid, row - 1, col);
    }
    // down
    if (row + 1 < row_num && grid[row + 1][col] == '1') {
      grid[row + 1][col] = '0';
      clearIslandDfsRecursive(grid, row + 1, col);
    }
    // left
    if (col - 1 >= 0 && grid[row][col - 1] == '1') {
      grid[row][col - 1] = '0';
      clearIslandDfsRecursive(grid, row, col - 1);
    }
    // right
    if (col + 1 < col_num && grid[row][col + 1] == '1') {
      grid[row][col + 1] = '0';
      clearIslandDfsRecursive(grid, row, col + 1);
    }
  }

 public:
  int numIslands(vector<vector<char>> &grid) {
    // grid is empty
    if (grid.size() == 0 || grid[0].size() == 0) {
      return 0;
    }

    int row_num = grid.size();
    int col_num = grid[0].size();
    int island_num = 0;

    for (int row = 0; row < row_num; ++row) {
      for (int col = 0; col < col_num; ++col) {
        if (grid[row][col] == '1') {
          island_num += 1;
          clearIslandDfsRecursive(grid, row, col);
        }
      }
    }

    return island_num;
  }
};

// @lc code=end
